Pie doughnut dataset weight

Vincent-Ip opened the pull request:
<!--
Please consider the following before submitting a pull request:

Guidelines for contributing: https://github.com/chartjs/Chart.js/blob/master/docs/developers/contributing.md

Example of changes on an interactive website such as the following:
- https://jsbin.com/
- https://jsfiddle.net/
- https://codepen.io/pen/
- Premade template: https://codepen.io/pen?template=JXVYzq
-->
I thought that this could help other people trying to adjust the relative size of Pie and Doughnut datasets.  I saw a feature request for something similar #4724 and I also had a need to vary the thickness to make the pie chart look they way I wanted.

Here's an example of what the chart looks like with the new dataset attribute - weight.
![screen shot 2019-01-01 at 11 52 09 pm](https://user-images.githubusercontent.com/40588938/50580989-3e194b00-0e23-11e9-80bb-f73921a5942c.png)

In the above example, the inner dataset was given a weight of 3.  This results in a data set that is 3 times the size of outer dataset (which doesn't specify a weight and so has a default weight of 1).

Fixes #4724
Fixes #5330
Fixes #5612

benmccann commented on src/controllers/controller.doughnut.js:
```diff
@@ -151,6 +159,7 @@ module.exports = DatasetController.extend({
 		var arcs = meta.data;
 		var cutoutPercentage = opts.cutoutPercentage;
 		var circumference = opts.circumference;
+		var chartWeight = (chart.data.datasets[me.index].weight) ? chart.data.datasets[me.index].weight : 1;
```
you don't need parens around `chart.data.datasets[me.index].weight`

benmccann commented on src/core/core.controller.js:
```diff
@@ -726,6 +726,20 @@ module.exports = function(Chart) {
 			return count;
 		},
 
+		getVisibleDatasetWeight: function() {
```
Can this be in the dougnut controller? It doesn't seem like something that all controllers need

Vincent-Ip commented:
@benmccann Thanks for the feedback.  I'll make those changes.

benmccann commented on src/controllers/controller.doughnut.js:
```diff
@@ -138,6 +138,37 @@ module.exports = DatasetController.extend({
 	},
 
+	// Get radius length offset of the dataset in relation to the visible datasets weights. This allows determining the inner and outer radius correctly
+	getRingWeightedOffset: function(datasetIndex) {
```
I would recommend making all these new methods private. I.e. prefix the name with `_` and mark as private in the jsdoc:
```
/**
 * @private
 */
```

See `_resolveElementOptions` in this file as an example.

benmccann commented on src/controllers/controller.doughnut.js:
```diff
@@ -138,6 +138,37 @@ module.exports = DatasetController.extend({
+		if (!ringWeight) {
+			ringWeight = 1;
+		}
+		return ringWeight;
```
you could rewrite this whole function as:

    return this.chart.data.datasets[dataSetIndex].weight || 1;

benmccann commented on src/controllers/controller.doughnut.js:
```diff
@@ -138,6 +138,37 @@ module.exports = DatasetController.extend({
+	getRingWeightedOffset: function(datasetIndex) {
+		var ringWeightOffset = 0;
+
+		for (var j = 0; j < datasetIndex; ++j) {
```
can you use `i` as the variable name since it's more standard

benmccann commented on src/controllers/controller.doughnut.js:
```diff
@@ -138,6 +138,37 @@ module.exports = DatasetController.extend({
+	},
+
+	getVisibleDatasetTotalWeight: function() {
+		var totalWeight = 0;
```
I think you could replace this entire function with:

    return getRingWeightedOffset(this.chart.data.datasets.length);

Vincent-Ip replied:
Thanks! I should have realized that one myself.

benmccann commented on src/controllers/controller.doughnut.js:
```diff
@@ -151,6 +151,7 @@ module.exports = DatasetController.extend({
 		var arcs = meta.data;
 		var cutoutPercentage = opts.cutoutPercentage;
 		var circumference = opts.circumference;
+		var chartWeight = chart.data.datasets[me.index].weight ? chart.data.datasets[me.index].weight : 1;
```
You could simplify this to:

    var chartWeight = chart.data.datasets[me.index].weight || 1

benmccann replied:
Or maybe `me._getRingWeight(me.index)`

nagix commented:
The `weight` values in datasets are referenced many times for a single `draw` call, so there is the same issue as the [comment](https://github.com/chartjs/Chart.js/pull/5841#discussion_r240979470) in #5841. I'm wondering if we can cache `weight` for each dataset and the total weight value when the first visible dataset is updated.

simonbrunel commented:
Good point @nagix but it may be better to tackle this optimization in a separate PR (maybe by registering an internal plugin computing these values during `beforeDatasetsUpdate`).

I like this feature, however we may want a different API for it. I was already thinking to use `weight` to control the dataset order when drawing ([many time requested](5171)) which I think matches better. But I also remember requests to [control the radius](3923) so we should think this API to support this case without duplicating options.

Highchart uses `size: '42%'`:
- `size: 42` would generate a ring with a size of 42px
- `size: "75%"` would be similar to this PR `weight: 3`

If we really want to implement the weight approach, maybe we could use the `flex` name (from CSS). But in this case, I'm not sure how it will work along the `size` (or whatever future option) controlling the size of the ring.

simonbrunel commented:
> I'm not sure how it will work along the size (or whatever future option) controlling the size of the ring.

`flex` could be ignored from all computations if `size` is explicitly defined. So I'm fine with both: `flex` (this PR) or `size` as percentage, however `flex` would be easier as a first step and already almost implemented. 

I would definitely reserve `weight` to control the dataset order, which applies to all type of charts. Also note that we already use [`weight` to sort scales](https://www.chartjs.org/docs/latest/axes/?h=weight), so would be more consistent. `flex` is pretty common, especially for [CSS developers](https://developer.mozilla.org/en-US/docs/Web/CSS/flex). Other suggestions for the option name are of course welcome :)

Thoughts?

Vincent-Ip commented:
Hi @simonbrunel , sorry for the slow reply.  I thought that perhaps you were asking the question of your co-reviewers.   I do agree with you that this parameter should be called something other than `weight`, since you are using `weight` in other places of something different.

`flex` seems like a possible fit, but the css usage of `flex` itself seems different (certainly at least in syntax).  Perhaps without knowing what the future plans for `size` and chart resizing behavior, i'm at a bit of a disadvantage for seeing how close this feature fits with the css usage of `flex` and how closely (or not), this feature would resemble it.

i'm happy to change it to `flex`, but i'm not sure it won't be confusing.   It will certainly be less confusing than `weight`.   In terms of other possible names for this feature, nothing exceptional comes to mind.  Perhaps it could be named something more explicit like `relative-size`, or `ring-scale`?

Sorry to give you back a question, but...  what do you think?

simonbrunel commented:
Thanks @Vincent-Ip 

I would go for a name that will also work for other charts, for example we could implement it to control the size of bars when side by side (which is of course not the purpose of this PR). That's why I would eliminate `ring-scale`. I would also discard `relative-size` because a `size` to me is `64px, 50%, 15rem` but not a `weight`.

We are not designing our features to support exactly the same syntax as CSS, I'm just saying that `flex` is a well-know CSS feature that represents exactly what this PR is trying to implement. 

For example:

```css
.col-1 { flex: 2; }  // size: 25%
.col-2 { flex: 2; }  // size: 25%
.col-3 { flex: 4; }  // size: 50%
``` 

In this case, it's the same syntax and the same behavior, so could be a good match.

benmccann commented:
Sorry to chime in a bit late. The name `flex` isn't very intuitive to me. It also had me initially a bit confused with the `barThickness: 'flex'` option that we have: https://www.chartjs.org/docs/latest/charts/bar.html#barthickness

It's been a few years since I've done much CSS. I'm guessing that's a newer addition since I wasn't familiar with it. The closest thing I could think of was HTML's `colspan`. Maybe we could call this `span`? Other naming ideas might be `proportion`, `magnitude`, `significance`. I also didn't think `weight` was a bad name and think it probably works better for this feature than for the order, but am fine picking something new

etimberg commented:
I'm fine with weight. It's what came to mind at first

etimberg commented:
Looks like this will also resolve #5330 #5612

kurkle commented:
> @kurkle I'm just looking at the screenshot you shared, can you post the associated pen?

https://codepen.io/kurkle/pen/XoxyrM

simonbrunel commented:
Thanks, that transparent dataset in the middle confused me a bit :)

nagix commented:
`wight` sounds more intutive to me while the term weight is very generic. How about `sizeWeight` / `sortWight`?

simonbrunel commented:
@nagix I guess you mean `weight` (not `wight`)? If `weight` is more intuitive for everyone (other than me), then I guess we got a winner, except if there is unanimity for `sizeWeight` :) (while I agree `weight` is very generic, I think I prefer `weight` instead of `sizeWeight`).

Then we can use something like `order` to sort the datasets array.

@Vincent-Ip sorry for this back and forth.

nagix commented on docs/charts/doughnut.md:
```diff
@@ -63,6 +63,7 @@ The doughnut/pie chart allows a number of properties to be specified for each da
 | [`hoverBackgroundColor`](#interations) | [`Color`](../general/colors.md) | Yes | Yes | `undefined`
 | [`hoverBorderColor`](#interactions) | [`Color`](../general/colors.md) | Yes | Yes | `undefined`
 | [`hoverBorderWidth`](#interactions) | `Number` | Yes | Yes | `undefined`
+| [`weight`](#styling) | `Number` | No | No | `1`
```
To be consistent, this should be 
```
| [`weight`](#styling) | `Number` | - | - | `1`
```

Vincent-Ip replied:
I'll look to include this change in the next update.  I would like to hear from you about the `weight: 0` question.

nagix commented:
`weight: 0` falls back to `1`, but I feel `0` is still a valid value (css `flex` accepts `0`). In case of pie/doughnut charts, thickness will be zero but the border could be drawn.

Vincent-Ip commented:
@nagix I hear what you are saying with a 0 thickness ring, but `weight: 0` could be problematic.  If `weight: 0` is allowed, someone could put all dataset weights to 0.  The code uses the total data set weight to judge proportions, and that could lead to a divide by 0 situation.  It is probably unlikely, but it would be technically possible.   

So with that little bit of information, should `weight: 0` be allowed?  If you think so, I can make some changes.

Vincent-Ip commented:
Thinking a little more, negative values would be problematic too.

Vincent-Ip commented:
Just doing some testing with a modified ChartJs 2.7.3 library with the code from this PR.  

https://codepen.io/vincent-ip/pen/PXLprq

I modified it to support `weight: 0`.  The behavior is a little funky, but I guess that what you should expect with `weight: 0`.  If you use multiple data sets, the borders will draw as expected, but they will overlap on top of each other.  I had to increase the `borderWidth` of the later datasets just so that you could see them.  They will also have no tooltip area, as you would guess.

**Just wanted to make sure this behavior was what you expected.**


----
Also in this pen, I adjusted to the code to handle the following: 
- Negative values are treated as `weight: 0`
- As for the total weight adding up to 0, I had to detect that scenario and return a total weight of 1.

etimberg commented:
re negative values, I am fine to fall back to `0`.

In terms of a `0` weight dataset, should the border be drawn or should it be treated as display none? I'm not sure what the most intuitive answer is. Based on flexbox I think we'd want it to draw the border but also move the other rings so that the borders do not fully overlap.

Vincent-Ip commented:
FYI, this last update addresses 
1. Negative `weight` values (and treats them as 0s)
2. Allows 0s for `weight` values (but still defaults an undefined dataset `weight` to 1)
3. Avoids a divide by 0 error when `_getVisibleDatasetWeightTotal` returns 0

What is not addressed is the overlap / hiding of `weight: 0` datasets.  Currently, `weight: 0` areas will have no tooltip area, but borders will still be drawn and will potentially still overlap other datasets. (as shown in this codepen. https://codepen.io/vincent-ip/pen/PXLprq)

Vincent-Ip commented:
Hi @nagix, i made those changes you suggested.  Is there any other changes that you think should be made?

nagix commented:
Sorry for the late response. Except for one review comment, looks good to me.

Vincent-Ip commented:
@nagix sure thing.  I've not used jasmine / karma before, but I can figure it out.

nagix commented on docs/charts/doughnut.md:
```diff
@@ -63,6 +63,7 @@ The doughnut/pie chart allows a number of properties to be specified for each da
 | [`hoverBackgroundColor`](#interations) | [`Color`](../general/colors.md) | Yes | Yes | `undefined`
 | [`hoverBorderColor`](#interactions) | [`Color`](../general/colors.md) | Yes | Yes | `undefined`
 | [`hoverBorderWidth`](#interactions) | `Number` | Yes | Yes | `undefined`
+| [`weight`](#styling) | `Number` | - | - | `1`
```
In #6020, all basic types were converted to lower case. So, `Number` should be `number`.

Vincent-Ip replied:
Thanks for the heads up.

Vincent-Ip commented:
Hi @nagix, 

I updated the file controller.doughnut.tests.js.  Is that correct / sufficient in terms of adding testing for this feature or is there more that needs to be done?

Thanks for your help.

simonbrunel commented on test/specs/controller.doughnut.tests.js:
```diff
@@ -325,6 +325,83 @@ describe('Chart.controllers.doughnut', function() {
 		expect(controller.innerRadius).toBe(126);
 	});
 
+	it('should create arcs with widths based on the weight value of the dataset', function() {
```
I think an image based test would have been better here, @nagix what do you think?

Vincent-Ip replied:
@nagix @simonbrunel Do you want to add an image based test?  I can look into it if need be,

simonbrunel replied:
I would replace these pixel based tests by image based tests since it's more robust and (IMO) easier to update. We are actually [writing some docs](https://github.com/chartjs/Chart.js/pull/6073) about this type of tests, you could give a try :)

simonbrunel replied:
@Vincent-Ip do you think we need to keep this test in `controller.doughnut.tests.js` since it's exactly the same as the fixture ones? Except if it tests something special, I think I would just remove it to avoid maintaining both later.

Vincent-Ip replied:
@simonbrunel Well, the pixel based tests do overlap with the image test, but the pixel tests are much more explicit, in terms of testing those newly added private methods and their resulting behavior.  Since these methods are private, I guess testing could be thought of as overkill, since no one is relying on their behavior, and the image based tests cover the general overall behavior.

I do want to say that there are similar existing methods that are public (and this code was originally written emulating existing public methods), so I could see that if any of these methods become part of a public api then it may call for the re-introduction of pixel based test coverage code.

But since these methods are currently private, we can let whatever future PR re-introduce pixel based tests, if needed.  **I will remove the pixel based tests**.  Apologies to all for the constant little updates and making you all have to re-approve this PR, but I guess that's just how the sausage is made.

giobi commented:
Is this gonna be merged in the next version?

benmccann commented on docs/charts/doughnut.md:
```diff
@@ -73,6 +74,7 @@ The style of each arc can be controlled with the following properties:
 | `backgroundColor` | arc background color.
 | `borderColor` | arc border color.
 | `borderWidth` | arc border width (in pixels).
+| `weight` | The relative thickness of the dataset. Providing a value for weight will allow the pie or doughnut dataset to be drawn with a thickness relative to the sum of all the dataset weight values.
```
I'd probably say "cause" instead of "allow"

Vincent-Ip replied:
thanks for the proof read

benmccann commented on src/controllers/controller.doughnut.js:
```diff
@@ -180,14 +182,14 @@ module.exports = DatasetController.extend({
-		me.outerRadius = chart.outerRadius - (chart.radiusLength * me.getRingIndex(me.index));
-		me.innerRadius = Math.max(me.outerRadius - chart.radiusLength, 0);
+		me.outerRadius = chart.outerRadius - (chart.radiusLength * me._getRingWeightOffset(me.index));
+		me.innerRadius = Math.max(me.outerRadius - (chart.radiusLength * chartWeight), 0);
```
I don't think the parentheses around `chart.radiusLength * chartWeight` are necessary

simonbrunel commented:
@Vincent-Ip I think we should remove the pixel based test since it's exactly the same as the image based ones (see [this comment](/chartjs/Chart.js/pull/5951/files#r2597115)).

simonbrunel commented:
Thanks @Vincent-Ip for this work and all the iterations.

Vincent-Ip commented:
@simonbrunel No problem.  Thank you and all other reviewers for the experience.

sheixt commented:
Ah looks like I missed the progression of this, was on following #4724. Looks great guys, I'll try to find the time to give it a spin and adjust my workaround 😄
